File: 2020/Python/Day13.py
import os
import sys


def main():
    inf = os.path.dirname(os.getcwd())
    inf = os.path.join(inf, "input")
    if not os.path.exists(inf) or os.path.isdir(inf):
        print(inf + " does not exist!")
        inf += ".txt"
        print(f"trying {inf} instead.")

    if not os.path.exists(inf) or os.path.isdir(inf):
        print(inf + " does not exist!")
        inf = os.path.join(os.path.dirname(inf), "input")
        inf = os.path.join(inf, "Day13.txt")
        print(f"trying {inf} instead.")

    if not os.path.exists(inf) or os.path.isdir(inf):
        print("None of the expected inputs exist!", file=sys.stderr)
        return
    else:
        print(f"Using input file {inf}.")

    inputFile = open(inf, 'r')
    departing = 0
    busIds = []
    busses = {}
    fastest = 0;
    for line in inputFile.readlines():
        if departing == 0:
            departing = int(line[:-1])
        else:
            for bus in line.split(','):
                if bus != "x":
                    b = int(bus)
                    busses[b] = b - departing % b
                    if fastest == 0 or busses[b] < busses[fastest]:
                        fastest = b
                    busIds.append(b)

                else:
                    busIds.append(0)

    print(f"Part one result is {fastest * busses[fastest]}.")

    time = 0
    step = 1
    current = -1
    while(True):
        bus = 0
        for i in range(current + 1, len(busIds)):
            bus = busIds[i]
            if bus == 0:
                continue
            elif (time + i) % bus != 0:
                bus = 0
                break;
            else:
                current = i
                step *= bus

        if bus == busIds[-1]:
            break
        else:
            time += step

    print(f"Part two timestamp is {time}.")


if __name__ == '__main__':
    main()

File: 2020/Java/src/aoc/Day13.java
package aoc;

import java.io.IOException;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public abstract class Day13 {

	public static void main(String[] args) throws IOException {
		int departing = 0;
		List<Integer> busIds = new ArrayList<>();
		Map<Integer, Integer> busses = new HashMap<>();
		int fastest = 0;
		for (String line : InputReader.readInputFileLines(13)) {
			if (departing == 0) {
				departing = Integer.parseInt(line);
			} else {
				for (String bus : line.split(",")) {
					if (!bus.equals("x")) {
						int b = Integer.parseInt(bus);
						busses.put(b, b - departing % b);
						if (fastest == 0 || busses.get(b) < busses.get(fastest)) {
							fastest = b;
						}
						busIds.add(b);
					} else
						busIds.add(0);
				}
			}
		}

		System.out.println("Part one result is " + fastest * busses.get(fastest) + ".");

		long time = 0;
		long step = 1;
		int current = -1;
		while (true) {
			int bus = 0;
			for (int i = current + 1; i < busIds.size(); i++) {
				bus = busIds.get(i);
				if (bus == 0) {
					continue;
				} else if ((time + i) % bus != 0) {
					bus = 0;
					break;
				} else {
					current = i;
					step *= bus;
				}
			}

			if (bus == busIds.get(busIds.size() - 1))
				break;
			else
				time += step;
		}

		System.out.println("Part two timestamp is " + time + ".");
	}
}
